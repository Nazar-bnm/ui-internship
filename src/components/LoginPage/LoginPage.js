import React from 'react';
import ls from 'local-storage';
import { Link } from 'react-router-dom';

import { Button } from '@/shared';
import HttpService from '../../service/HttpService/httpService';
import { VALIDATION_FAILED, LOGIN_FAILED, LOGIN_SUCCESS } from '../../constants/notificationData';

import './LoginPage.scss';

const CN = 'login-page';

class LoginPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      emailToLogin: '',
      passwordToLogin: '',
      emailToRegister: ''
    };

    this.formLogin = React.createRef();
    this.formRegister = React.createRef();
    this.validate = this.validate.bind(this);
  }

  componentDidMount() {
    this.setState({
      emailToLogin: ls.get('emailToLogin') || '',
      passwordToLogin: ls.get('passwordToLogin') || '',
      emailToRegister: ls.get('emailToRegister') || ''
    });
  }

  saveToLocalStorage() {
    const emailToLoginValue = this.formLogin.current[0].value;
    const passwordToLoginValue = this.formLogin.current[1].value;
    const emailToRegisterValue = this.formRegister.current[0].value;

    ls.set('emailToLogin', emailToLoginValue);
    ls.set('passwordToLogin', passwordToLoginValue);
    ls.set('emailToRegister', emailToRegisterValue);
  }

  async doesUserExist() {
    const { showMessage } = this.props;
    const userAPI = new HttpService();
    const usersURL = `${process.env.SERVER_URL}/login`;
    const emailToLoginValue = this.formLogin.current[0].value;
    const passwordToLoginValue = this.formLogin.current[1].value;
    const currentUser = {
      email: emailToLoginValue,
      password: passwordToLoginValue
    };

    try {
      const response = await userAPI.post(usersURL, currentUser);
      const errorMessage = response.data.message;

      if (errorMessage) {
        showMessage(LOGIN_FAILED);
      } else {
        ls.set('authToken', response.data.token);
        showMessage(LOGIN_SUCCESS);
        window.history.back();
      }
    } catch (error) {
      showMessage(LOGIN_FAILED);
    }
  }

  validate({ target }) {
    const { showMessage } = this.props;
    const clickedButton = target.innerText.toUpperCase();
    let isInputValid = false;

    isInputValid = (clickedButton === 'LOGIN' ? this.formLogin : this.formRegister).current.reportValidity();

    if (!isInputValid) {
      showMessage(VALIDATION_FAILED);
    } else {
      this.saveToLocalStorage();
      this.doesUserExist();
    }

    return isInputValid;
  }

  render() {
    const { emailToLogin, passwordToLogin, emailToRegister } = this.state;

    return (
      <div className={`${CN}__wrapper content`}>
        <h3 className={`${CN}__main-heading`}>user login</h3>
        <div className={`${CN}__login-register-wrapper`}>
          <div className={`${CN}__registration-type-wrapper`}>
            <h3 className={`${CN}__heading`}>Registered customers</h3>
            <p className={`${CN}__text`}>
              <strong>Already registered? </strong>
              Please login below:
            </p>
            <form
              className={`${CN}__inputs-wrapper`}
              ref={this.formLogin}
              onSubmit={(e) => e.preventDefault()}
            >
              <span className={`${CN}__label`}>e-mail*</span>
              <br />
              <input
                type="email"
                name="email"
                className={`${CN}__input`}
                defaultValue={emailToLogin}
                required
              />
              <br />
              <span className={`${CN}__label`}>password*</span>
              <br />
              <input
                type="password"
                name="password"
                className={`${CN}__input`}
                defaultValue={passwordToLogin}
                required
                pattern="^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$"
              />
              <br />
            </form>
            <div className={`${CN}__button-and-link-wrapper`}>
              <Button
                className="black-button"
                onClick={this.validate}
              >
                Login
              </Button>
              <a
                className={`${CN}__link`}
                href="/"
                onClick={(e) => e.preventDefault()}
              >
                Forgor your password?
              </a>
            </div>
          </div>
          <div className={`${CN}__registration-type-wrapper`}>
            <h3 className={`${CN}__heading`}>New customers</h3>
            <p className={`${CN}__text`}>
              <strong>
                Enter your email address to create an account:
              </strong>
            </p>
            <form
              className={`${CN}__new-customer-input-wrapper`}
              ref={this.formRegister}
            >
              <span className={`${CN}__label`}>e-mail*</span>
              <br />
              <input
                type="email"
                name="email"
                className={`${CN}__input`}
                defaultValue={emailToRegister}
                required
              />
            </form>
            <Link to="/RegisterForm">
              <Button
                className="black-button"
                onClick={this.validate}
              >
              create an account
              </Button>
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

export default LoginPage;
