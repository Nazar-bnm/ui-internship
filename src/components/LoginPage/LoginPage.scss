@import '../../scss/global_styles';

.login-page {
  &__wrapper {
    margin-top: 20px;
    border: 1px solid $silver-chalice;
  }

  &__main-heading {
    padding: 10px;
    border-bottom: 1px solid $silver-chalice;
    text-transform: uppercase;
  }

  &__heading {
    padding: 10px 0;
    text-transform: uppercase;
  }

    &__login-register-wrapper {
      @include flex;
      padding: 4vw;
      font-family: font(primary-light);
    }
  
    &__registration-type-wrapper {
      @include flex($direction: column, $justify: space-between);
      width: 280px;

      &:nth-of-type(2) {
        justify-content: flex-start;
      }
    }

    &__new-customer-input-wrapper {
      margin-bottom: 10px;
      width: 100%;
    }
  
    &__label {
      text-transform: uppercase;
    }
  
    &__input {
      padding: 5px;
      margin: 5px 0;
      width: 100%;
      border: 1px solid $silver-chalice;
    }
  
    &__link {
      margin-top: 10px;
      position: relative;
  
      &:hover:after{
        content: 'Please contact our support team to restore your password.';
        position: absolute;
        left: 0;
        bottom: 20px;
        min-width: 200px;
        border: 1px $dove-gray solid;
        border-radius: 10px;
        background-color: $mercury;
        padding: 12px;
        color: #000000;
        font-size: 14px;
        z-index: 1;
      }
    }
  
    &__heading {
      margin: 10px 0;
      text-transform: uppercase;
    }
  
    &__text {
      margin-bottom: 10px;
    }
  
    &__button-and-link-wrapper {
      @include flex($justify: space-between, $align: center);
      margin-top: 10px;
    }
  }
  
  @media only screen and (min-width: 480px) {
    .login-page {
      &__main-heading {
        text-align: center;
      }

      &__login-register-wrapper {
        @include flex($justify: space-between);
  
        &__input, &__button-and-link-wrapper {
          width: 200px;
        }

        &__new-customer-input-wrapper {
          margin-bottom: 0;
        }
    
        &__registration-type-wrapper {
          width: 45%;

          &:nth-of-type(2) {
            .black-button {
              width: 300px;
            }
          }
        }
      }
    }
  }

  @media only screen and (min-width: 768px) {
    .login-page {
      &__registration-type-wrapper {
        &:nth-of-type(2) {
          .black-button {
            width: 100%;
          }
        }
      }
    }
  }

  @media only screen and (min-width: 1024px) {
    .login-page {
      &__input, &__button-and-link-wrapper {
        width: 300px;
      }

      &__registration-type-wrapper {
        width: 35vw;

        &:nth-of-type(2) {
          .black-button {
            width: 300px;
          }
        }
      }
    }
  }