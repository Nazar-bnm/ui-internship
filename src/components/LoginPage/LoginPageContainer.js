import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import LoginPage from './LoginPage';
import showMessage from '../../actions/actionsNotifications';

const mapDispatchToProps = (dispatch) => ({
  showMessage: bindActionCreators(showMessage, dispatch)
});

export default connect(null, mapDispatchToProps)(LoginPage);
