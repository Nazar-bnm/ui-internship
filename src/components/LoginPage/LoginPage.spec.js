import React from 'react';
import { shallow } from 'enzyme';

import LoginPage from './LoginPage';

describe('<LoginPage />', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
  });
});
