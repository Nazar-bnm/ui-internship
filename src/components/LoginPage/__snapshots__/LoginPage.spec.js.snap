// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should match snapshot 1`] = `
<div
  className="login-page__wrapper content"
>
  <h3
    className="login-page__main-heading"
  >
    user login
  </h3>
  <div
    className="login-page__login-register-wrapper"
  >
    <div
      className="login-page__registration-type-wrapper"
    >
      <h3
        className="login-page__heading"
      >
        Registered customers
      </h3>
      <p
        className="login-page__text"
      >
        <strong>
          Already registered? 
        </strong>
        Please login below:
      </p>
      <form
        className="login-page__inputs-wrapper"
        onSubmit={[Function]}
      >
        <span
          className="login-page__label"
        >
          e-mail*
        </span>
        <br />
        <input
          className="login-page__input"
          defaultValue=""
          name="email"
          required={true}
          type="email"
        />
        <br />
        <span
          className="login-page__label"
        >
          password*
        </span>
        <br />
        <input
          className="login-page__input"
          defaultValue=""
          name="password"
          pattern="^(?=.*[A-Za-z])(?=.*\\\\d)[A-Za-z\\\\d]{8,}$"
          required={true}
          type="password"
        />
        <br />
      </form>
      <div
        className="login-page__button-and-link-wrapper"
      >
        <Button
          className="black-button"
          icon={null}
          onClick={[Function]}
        >
          Login
        </Button>
        <a
          className="login-page__link"
          href="/"
          onClick={[Function]}
        >
          Forgor your password?
        </a>
      </div>
    </div>
    <div
      className="login-page__registration-type-wrapper"
    >
      <h3
        className="login-page__heading"
      >
        New customers
      </h3>
      <p
        className="login-page__text"
      >
        <strong>
          Enter your email address to create an account:
        </strong>
      </p>
      <form
        className="login-page__new-customer-input-wrapper"
      >
        <span
          className="login-page__label"
        >
          e-mail*
        </span>
        <br />
        <input
          className="login-page__input"
          defaultValue=""
          name="email"
          required={true}
          type="email"
        />
      </form>
      <Button
        className="black-button"
        icon={null}
        onClick={[Function]}
      >
        create an account
      </Button>
    </div>
  </div>
</div>
`;
